### PR TITLE
Add depth support

### PR DESCRIPTION
**Pull request by stojicnnnn**
- Added cameras/realsense.py because of changes made to async_read from lerobot
- Changed infere and alignnet to accept both rgb and depth
- Multiplied rotation part of relative_action matrix to speed up RPY alignment
- Added new trajectory generation functions for data collection
- Visualisation now shows both rgb and depth images
- Added depth testing to tests
- Fixed formatting issues
- Depth clipping done before train and infere, makes sure that the input to network is always the same
- Fixed Z offset after infere

### CODE REVIEW

File: cameras/realsense.py [hunk 1]
```diff
@@ -0,0 +1,552 @@
+#reference to https://github.com/huggingface/lerobot/tree/main
+#async_read now returns both rgb and depth images
+
+
+"""
+Provides the RealSenseCamera class for capturing frames from Intel RealSense cameras.
+"""
+
```
Comment (lukicdarkoo) [inline cameras/realsense.py:8]:
```suggestion
# Reference to https://github.com/huggingface/lerobot/tree/main
# Changed async_read now returns both RGB and depth images
"""

```

Comment (lukicdarkoo) [reply cameras/realsense.py:8]: Also, the folder is wrong. Please test on the hardware

File: alignit/models/alignnet.py [hunk 2]
```diff
@@ -7,11 +7,13 @@
 class AlignNet(nn.Module):
     def __init__(
         self,
-        backbone_name="efficientnet_b0",
+        backbone_name="efficientnet_b0",
```
Comment (lukicdarkoo) [inline alignit/models/alignnet.py:10]:
```suggestion
        backbone_name="efficientnet_b0",
```
Unnecessary space

File: alignit/models/alignnet.py [hunk 3]
```diff
@@ -7,11 +7,13 @@
 class AlignNet(nn.Module):
     def __init__(
         self,
-        backbone_name="efficientnet_b0",
+        backbone_name="efficientnet_b0", 
         backbone_weights="DEFAULT",
         use_vector_input=True,
+        use_depth_input=True,  # NEW
```
Comment (lukicdarkoo) [inline alignit/models/alignnet.py:13]:
```suggestion
        use_depth_input=True,
```
Misleading comment

File: alignit/models/alignnet.py [hunk 4]
```diff
@@ -7,11 +7,13 @@
...
         fc_layers=[256, 128],
         vector_hidden_dim=64,
+        depth_hidden_dim=128,  # NEW
```
Comment (lukicdarkoo) [inline alignit/models/alignnet.py:16]:
```suggestion
        depth_hidden_dim=128,
```
Misleading comment

File: alignit/robots/xarm.py [hunk 5]
```diff
@@ -33,10 +33,12 @@ def send_action(self, action):
         self.robot.send_action(action)
 
     def get_observation(self):
-        rgb_image = self.camera.read()
+        rgb_image,depth_image,acqusition_time = self.camera.async_read()
```
Comment (lukicdarkoo) [inline alignit/robots/xarm.py:36]:
```suggestion
        rgb_image, depth_image, acquisition_time = self.camera.async_read()
```
Space after comma, `acquisition` misspelled

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
Looks good overall, but please carefully check the diff

File: alignit/train.py [hunk 6]
```diff
@@ -52,35 +46,47 @@ def main(cfg: TrainConfig):
...
+                if depth_images_pil[i] is not None: # Check if depth is actually present for this item
+                    depth_sequence = depth_images_pil[i]
+                    tensor_sequence_depth = [
+                        depth_transform(d_img.convert("L")) for d_img in depth_sequence
+                    ] 
+                    stacked_tensors_depth = torch.stack(tensor_sequence_depth, dim=0)
+                    batch_depth_tensors.append(stacked_tensors_depth)
+                else:
+                    print(f"Warning: Depth image missing for sample {i} in batch.")
+                    raise ValueError("Depth images expected but not found for a sample when use_depth is True.")
```
Comment (lukicdarkoo) [inline alignit/train.py:78]: Maybe I am missing, but I don't see the `use_depth` check

File: alignit/config.py [hunk 7]
```diff
@@ -149,6 +156,11 @@ class InferConfig:
         default=5,
         metadata={"help": "Number of iterations within tolerance before stopping"},
     )
+    num_multiplications: int = field(
+        default=3,
+        metadata={
+            "help": "Number of times to multiply the transformation matrix"},
```
Comment (lukicdarkoo) [inline alignit/config.py:162]: This is not the multiplication of transformation matrices but rotation matrices.

Also, the parameter name and description are too generic.

File: alignit/train.py [hunk 8]
```diff
@@ -52,35 +46,47 @@ def main(cfg: TrainConfig):
...
+                stacked_tensors_rgb = torch.stack(tensor_sequence_rgb, dim=0)
+                batch_rgb_tensors.append(stacked_tensors_rgb)
 
-            # Stack all batches to get shape (B, N, 3, H, W)
-            batch_images = torch.stack(batch_images, dim=0).to(device)
+                # Process Depth
+                if depth_images_pil[i] is not None: # Check if depth is actually present for this item
+                    depth_sequence = depth_images_pil[i]
+                    tensor_sequence_depth = [
+                        depth_transform(d_img.convert("L")) for d_img in depth_sequence
```
Comment (lukicdarkoo) [inline alignit/train.py:72]: We have to check this line, which will convert the depth image to 8-bit. I don't think it normalizes the depth image to the 0-1 range.

If this is the case, then we don't normalize images in the same way during inference and training, which is critical for performance.

**Review by lukicdarkoo — CHANGES_REQUESTED** [review]
